src/key_packages/mod.rs

```diff
@@ -274,22 +275,69 @@ impl KeyPackageBundle {
     pub fn new_with_keypair(
         ciphersuites: &[CiphersuiteName],
         credential_bundle: &CredentialBundle,
-        extensions: Vec<Box<dyn Extension>>,
+        mut extensions: Vec<Box<dyn Extension>>,
         key_pair: HPKEKeyPair,
         leaf_secret: Vec<u8>,
     ) -> Result<Self, ConfigError> {
         debug_assert!(!ciphersuites.is_empty());
-        let capabilities_extension = CapabilitiesExtension::new(None, Some(ciphersuites), None);
-        let mut final_extensions: Vec<Box<dyn Extension>> = vec![Box::new(capabilities_extension)];
 
+        // Check if the extensions contain any duplicates.
+        if (1..extensions.len()).any(|i| extensions[i..].contains(&extensions[i - 1])) {
+            return Err(ConfigError::DuplicateExtension);
+        }
+        //for ext in &extensions {
+        //    if extensions.iter().any(|e|  e.get_type() == ext.get_type()) {
+        //        return Err(ConfigError::DuplicateExtension);
+        //    }
+        //}
+
+        // Check if the `extensions` already contain the mandatory extensions
+        // and in case one of them is a capabilities extension, add the input
```

Comment:
Would it then not be easier to just remove the `ciphersuites` from the API and just read the ciphersuites off the capabilities extension? And if we have none, we can still take the default one...

Reply:
Sure that's an option as well as long as the capabilities extension is easy enough to assemble. I can see that ciphersuites is for example one people might want to change often but leave other capabilities as defaults. Maybe the capabilities extension should have a builder pattern for creation or allow modifying parts.

Reply:
That would be handy. You had already suggested that you wanted to rewrite Extensions at some point. Maybe this would be one issue to tackle at that point? For now, I'll just do what you propose initially and fail if the capabilities and the ciphersuites don't match.